### PR TITLE
#1027: Create Topic Cache for use in frontend homepage controller

### PR DESCRIPTION
**Pull request by joegasewicz**
### What

- Use dp-cache library to handle caching of data, you will need to abide by the cache interface and create your on methods to get data from Topics API
  - Documentation in dp-cache library for instructions on how to build your specific cache

- Initial startup needs to call topics API to get data and update topics cache

- Cache is updated by calling topics API every 30minutes
  - this should be using go routines like other examples in the homepage or search controller
  - cache interval should be configurable
  - when calling topics API we should be going through the api router, see application architecture in notes

- Check and extend registered checkers with topics API (I think this is there already due to navbar work however this maybe wrapped in if statements based on navbar feature being turned on or off)
  - Use the new feature flag in combination with existing flags to decide whether the topic API should be registered in health checks or not

- Use the new feature flag to determine whether the cache is instantiated or not (no point spinning up go routine if cache is not going to be used due to feature being turned off)

### Who can review
Anyone

### CODE REVIEW

File: cache/private/census_topic.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		} else {
+			err := errors.New("root topic public items is nil")
+			log.Error(ctx, "failed to deference root topics items pointer", err)
+			return cache.GetEmptyCensusTopic()
+		}
+
+		var censusTopicCache *cache.Topic
+
+		// go through each root topic, find census topic and gets its data for caching which includes subtopic ids
+		for i := range rootTopicItems {
```
Comment (nshumoogum) [inline cache/private/census_topic.go:44]: Do we need to call root topic to get a list of root topics and the loop through topics to find census id, we already know the id so we can use the `cache.CensusTopicID` to call it's subtopics?

File: cache/public/census_topic.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+// UpdateCensusTopic is a function to update the census topic cache in web (public) mode.
+// This function talks to the dp-topic-api via its public endpoints to retrieve the census topic and its subtopic ids
+// The data returned by the dp-topic-api is of type *models.PublicSubtopics which is then transformed to *cache.Topic in this function for the controller
+// If an error has occurred, this is captured in log.Error and then an empty census topic is returned
+func UpdateCensusTopic(ctx context.Context, topicClient topicCli.Clienter) func() *cache.Topic {
+	return func() *cache.Topic {
+		// get root topics from dp-topic-api
+		rootTopics, err := topicClient.GetRootTopicsPublic(ctx, topicCli.Headers{})
```
Comment (nshumoogum) [inline cache/public/census_topic.go:22]: Again not sure if this call is needed, as we know what the census topic id is

File: cache/subtopic.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"strings"
+	"sync"
+)
+
+// SubtopicsIDs contains a list of subtopics in map form with addition to mutex locking
+// The subtopicsMap is used to keep a record of subtopics to be later used to generate the subtopics id `query` for a topic
+// and to check if the subtopic id given by a user exists
+type SubtopicsIDs struct {
+	mutex        *sync.RWMutex
+	subtopicsMap map[string]bool
```
Comment (nshumoogum) [inline cache/subtopic.go:13]: So we will need the subtopic name/label and release date as well as the topic ids; these need to be stored in the cache for use by the census hub page view/template

Comment (nshumoogum) [reply cache/subtopic.go:13]: Root topic (census) shouldn't be included

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: cache/public/census_topic.go (hunk 2 above)
Comment (joegasewicz) [reply cache/public/census_topic.go:22]: see comment above please

**Review by nshumoogum — DISMISSED** [review]

File: cache/private/census_topic.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+		if censusSubtopics.PrivateItems == nil {
+			err := errors.New("root topic public items is nil")
+			log.Error(ctx, "failed to deference root topics items pointer", err)
+			return cache.GetEmptyCensusTopic()
+		}
+
+		var censusTopicCache *cache.Topic
+
+		censusTopicCache = setTopicCachePublic(ctx, *censusTopic, *censusSubtopics.PrivateItems)
```
Comment (hareshsinghkainth) [inline cache/private/census_topic.go:50]: Combine declaration and initliation into a single statement i.e. censusTopicCache := setTopicCachePublic(...)

File: cache/public/census_topic.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+		if censusSubtopics.PublicItems == nil {
+			err := errors.New("root topic public items is nil")
+			log.Error(ctx, "failed to deference root topics items pointer", err)
+			return cache.GetEmptyCensusTopic()
+		}
+
+		var censusTopicCache *cache.Topic
+
+		censusTopicCache = setTopicCachePublic(ctx, *censusTopic, *censusSubtopics.PublicItems)
```
Comment (hareshsinghkainth) [inline cache/public/census_topic.go:50]: Combine declaration and initalisation into a single statement i.e. censusTopicCache := setTopicCachePublic(...)

File: cache/public/census_topic.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+func setTopicCachePublic(ctx context.Context, censusTopic models.Topic, subtopics []models.Topic) *cache.Topic {
+	censusTopicCache := &cache.Topic{
+		ID:              censusTopic.ID,
+		LocaliseKeyName: censusTopic.Title,
+	}
+
+	subtopicCache := cache.NewSubTopicsMap()
+	for i := range subtopics {
+		subtopicCache.AppendSubtopicItems(subtopics[i])
```
Comment (hareshsinghkainth) [inline cache/public/census_topic.go:70]: Better to use the second iteration (value) variable from the for-range i.e. for _, value := range subtopics {...}

File: cache/subtopic.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+	"github.com/ONSdigital/dp-topic-api/models"
+)
+
+// SubtopicsIDs contains a list of subtopics in map form with addition to mutex locking
+// The subtopicsMap is used to keep a record of subtopics to be later used to generate the subtopics id `query` for a topic
+// and to check if the subtopic id given by a user exists
+type SubtopicsIDs struct {
+	mutex        *sync.RWMutex
+	subtopicsMap map[string]*models.Topic
```
Comment (hareshsinghkainth) [inline cache/subtopic.go:15]: Prefer to use sync.Map

File: cache/topic.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		return getEmptyTopic(), err
+	}
+
+	return topicCacheData, nil
+}
+
+// AddUpdateFunc adds an update function to the topic cache for a topic with the `title` passed to the function
+// This update function will then be triggered once or at every fixed interval as per the prior setup of the TopicCache
+func (dc *TopicCache) AddUpdateFunc(id string, updateFunc func() *Topic) {
+	dc.UpdateFuncs[id] = func() (interface{}, error) {
```
Comment (hareshsinghkainth) [inline cache/topic.go:81]: More defensive programming but any validation on id argument?

File: service/service.go [hunk 9]
```diff
@@ -37,26 +41,32 @@ func (svc *Service) Run(ctx context.Context, cfg *config.Config, serviceList *Ex
 
 	// Initialise render client
 	rend := render.NewWithDefaultClient(assets.Asset, assets.AssetNames, cfg.PatternLibraryAssetsPath, cfg.SiteDomain)
-
+	log.Info(ctx, "got here 1")
```
Comment (hareshsinghkainth) [inline service/service.go:44]: Do we need this? If not, please remove it.

File: service/service.go [hunk 10]
```diff
@@ -37,26 +41,32 @@ func (svc *Service) Run(ctx context.Context, cfg *config.Config, serviceList *Ex
...
 	routes.Init(
 		ctx,
 		r,
 		cfg,
+		svc.Cache,
 		svc.HealthCheck.Handler,
 		svc.HomePageClient,
 		rend,
 	)
+	log.Info(ctx, "got here 2")
```
Comment (hareshsinghkainth) [inline service/service.go:56]: Do we need this? If not, please remove it.

File: service/service.go [hunk 11]
```diff
@@ -37,26 +41,32 @@ func (svc *Service) Run(ctx context.Context, cfg *config.Config, serviceList *Ex
...
 	svc.Server = serviceList.GetHTTPServer(cfg.BindAddr, r)
+	log.Info(ctx, "got here 3")
```
Comment (hareshsinghkainth) [inline service/service.go:58]: Do we need this? If not, please remove it.

File: service/service.go [hunk 12]
```diff
@@ -37,26 +41,32 @@ func (svc *Service) Run(ctx context.Context, cfg *config.Config, serviceList *Ex
...
 
 	// Start Healthcheck and HTTP Server
 	svc.HealthCheck.Start(ctx)
+	log.Info(ctx, "got here 4")
```
Comment (hareshsinghkainth) [inline service/service.go:62]: Do we need this? If not, please remove it.

File: service/service.go [hunk 13]
```diff
@@ -37,26 +41,32 @@ func (svc *Service) Run(ctx context.Context, cfg *config.Config, serviceList *Ex
...
 	go func() {
 		if err := svc.Server.ListenAndServe(); err != nil {
 			svcErrors <- errors.Wrap(err, "failure in http listen and serve")
 		}
 	}()
+
+	log.Info(ctx, "got here 5")
```
Comment (hareshsinghkainth) [inline service/service.go:69]: Do we need this? If not, please remove it.

**Review by hareshsinghkainth — CHANGES_REQUESTED** [review]
A few minor comments.

**Review by hareshkainthmethods — APPROVED** [review]
LGTM

**Pull request merged**